### PR TITLE
Map class implementation

### LINKED ISSUE #4: Implement Map class Task#4

### PR DESCRIPTION
**Pull request by J-Cieplinski**
Closes #4

### CODE REVIEW

**Review by ArturGubala — APPROVED** [review]

**Review by iprice73 — APPROVED** [review]

File: shm/Map.hpp
Comment (niemarcin) [inline shm/Map.hpp:7]: I think we should decide on particular style of private/public/protected blocks of code. In Cargo class I've done public first, as it was in example solution from the lesson. Do you know what are good/bad practices in this case?

I found this very intuitive, but it's just stackoverflow: [https://stackoverflow.com/a/1760867/8964273](url)

> Personally, I prefer putting public methods first, followed by protected methods, following by private methods. Member data should in general always be private or protected, unless you have a good reason for it not to be so.
> 
> My rationale for putting public methods at the top is that it defines the interface for your class, so anyone perusing your header file should be able to see this information immediately.
> 
> In general, private and protected members are less important to most people looking at the header file, unless they are considering modifying the internals of the class. Keeping them "out of the way" ensures this information is maintained only on a need to know basis, one of the more important aspects of encapsulation.

File: shm/Map.hpp
Comment (niemarcin) [inline shm/Map.hpp]: Do we need {} for this pointer initialization?

File: shm/Map.hpp
Comment (niemarcin) [inline shm/Map.hpp:12]: Actually these functions were not required in Task4, but it's probably good to have them.

**Review by niemarcin — COMMENTED** [review]
@PierwszyKrytykRP please take a look at my comments before approval.

File: shm/Map.hpp
Comment (J-Cieplinski) [reply shm/Map.hpp:7]: TBH I do private fields first because usually I know what private members class will have at the beginning and  they are needed later for public methods etc. It's not very intuitive for me to write declaration of public getter for member `size_` when private variable `size_` wasn't even defined yet

File: shm/Map.hpp
Comment (J-Cieplinski) [reply shm/Map.hpp]: So I've initially thought that empty initializer list would initialize the pointer with nullptr. Turns out that's not the case. It is initialized by some null value. I've changed to reflect what I wanted it to do.

**Review by lisie-sprawy — APPROVED** [review]

File: shm/Map.hpp
Comment (J-Cieplinski) [reply shm/Map.hpp:12]: Yea I got ahead of myself. I'm sure they will be needed at some point in this or slightly changed form

**Review by niemarcin — APPROVED** [review]

**Pull request merged**
